### PR TITLE
Major changes and improvements

### PR DESCRIPTION
**Pull request by itsMarcoSolis**
Added create folder button, implemented AppBar for better visibility, changed selection button, renamed user variables to avoid confusion and implemented a better way to prevent user from accessing forbidden folders

### CODE REVIEW

File: lib/src/pick.dart [hunk 1]
```diff
@@ -8,28 +8,34 @@ import 'ls.dart';
 typedef Future<void> AfterPickedAction(BuildContext context, Directory folder);
 
 class _FolderPickerState extends State<FolderPicker> {
-  _FolderPickerState({@required this.action, @required this.rootDirectory}) {
-    _currentDirectory ??= rootDirectory;
```
Comment (itsMarcoSolis) [inline lib/src/pick.dart:12]: Root directory now refers to the actual root directory, whose parent can't be accessed by the user. When the state is created, it checks if the user has permissions to access the current folder's parent, if not, the current folder becomes now the root folder, preventing the user from going up on the folder tree.

File: lib/src/pick.dart [hunk 2]
```diff
@@ -40,38 +46,88 @@ class _FolderPickerState extends State<FolderPicker> {
...
-                GestureDetector(
-                  child: ListTile(
-                    leading: const Icon(Icons.arrow_upward),
-                    title: const Text("..", textScaleFactor: 1.5),
-                  ),
-                  onTap: () => setState(
-                      () => _currentDirectory = _currentDirectory.parent),
-                )
-              ]
-          ..addAll(_getDirs(context)),
```
Comment (itsMarcoSolis) [inline lib/src/pick.dart:57]: Every time the user goes up on the folder tree, we check if the upper folder is accesible, if it's not, now the current folder is the root folder and the user won't be able to go up, thus preventing the "Acces Denied" red screen.

File: README.md [hunk 3]
```diff
@@ -24,4 +24,5 @@ A directory picker for Flutter
 
    ```xml
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
```
Comment (itsMarcoSolis) [inline README.md:27]: The create new folder button needs writing permissions.

File: lib/src/pick.dart [hunk 4]
```diff
@@ -8,28 +8,34 @@ import 'ls.dart';
...
               icon: const Icon(Icons.folder, color: Colors.yellow),
-              onPressed: () => setState(() => _currentDirectory = dir)),
+              onPressed: () => setState(() => currentDirectory = dir)),
           title: GestureDetector(
-              onTap: () => setState(() => _currentDirectory = dir),
+              onTap: () => setState(() => currentDirectory = dir),
               child: Text(basename(dir.path))),
           trailing: IconButton(
-              icon: const Icon(Icons.file_download),
+              icon: const Icon(Icons.check_circle),
```
Comment (itsMarcoSolis) [inline lib/src/pick.dart:38]: Icon changed, just a personal preference.

File: lib/src/pick.dart [hunk 5]
```diff
@@ -8,28 +8,34 @@ import 'ls.dart';
...
   }
 
+  final textControllerName = new TextEditingController(text: "New Folder");
   final AfterPickedAction action;
-  final Directory rootDirectory;
+  Directory rootDirectory;
 
-  Directory _currentDirectory;
+  Directory currentDirectory;
+  Widget bodyList;
```
Comment (itsMarcoSolis) [inline lib/src/pick.dart:24]: A variable to make it easy to manage the folder list state.

**Comment by synw** [comment]
Hi. Thanks for the PR. 

- About the appbar: this FolderPicker widget is just a component that can be embeded not only in a full page but anywhere, and if the user wants an appbar he can manage. That said we can provide a separate widget like FolderPickerPage

- About the add folder button: this should be an option I think: maybe using a separated widget for this would be better as well

I tried to run the example from your fork and:

- There is a breaking change: the `rootFolder` parameter was replaced with `currentFolder`. I would prefer no breaking change. Why not to keep the paradigm of the root folder defined by the user?

- A crash occurs when running it: it says that the `dir` parameter of `ls` is null

- About parent folders permission check yes this is a good idea.

Can you provide a working example? Note: Unfortunately I committed a fix for the example permission problem and did some small improvements just a moment before your PR

**Comment by synw** [comment]
By the way I was thinking of a major refactor using https://github.com/synw/filex. It already has a controller for the add/delete folder actions and is quite configurable. This would reduce the maintenance here. This refactor could be a good occasion to integrate your changes

**Comment by synw** [comment]
@mocomaco hi, I made the refactor using the Filex package and it will simplify maintenance and avoid duplicate code. I added features based on the changes you wanted to implement:

- Ability to create a directory
- A ready to use page with an appbar showing the current folder name
- You can now choose you icon: I used the one you suggest for default

About permissions: the user can not go upper the rootDirectory. I have think about making the permission check inside the module but it is not compatible with desktop (I use this module in desktop apps) so it will stay outside.

Thanks a lot for your work and sorry about the refactoring but it had to be done. Tell me if everything you wanted is here now.
If you want any feature or have remarks you are welcome to open issues. Note for the future: it would be better to discuss the features in issues before making a PR to stay synchronized and avoid unnecessary work

**Pull request closed without merging**